sqlmesh/core/notification_target.py

```diff
@@ -43,20 +52,116 @@ def is_progress(self) -> bool:
+    AUDIT_FAILURE = "audit_failure"
+
+
+def notify(event: NotificationEvent) -> t.Callable:
+    """Decorator used to register 'notify' methods and the events they correspond to."""
+
+    def decorator(f: t.Callable) -> t.Callable:
+        @functools.wraps(f)
+        def wrapper(*args: t.List[t.Any], **kwargs: t.Dict[str, t.Any]) -> None:
+            return f(*args, **kwargs)
+
+        NOTIFICATION_FUNCTIONS[event] = f.__name__
+        return wrapper
+
+    return decorator
+
+
+class BaseNotificationTarget(PydanticModel, frozen=True):
     """
     Base notification target model. Provides a command for sending notifications that is currently only used
     by the built-in scheduler. Other schedulers like Airflow use the configuration of the target itself
     to create the notification constructs appropriate for the scheduler.
     """
 
     type_: str
+    notify_on: t.FrozenSet[NotificationEvent] = frozenset()
 
     def send(self, notification_status: NotificationStatus, msg: str, **kwargs: t.Any) -> None:
         """
-        Sends notification with the provided message. Currently only used by the built-in scheduler.
+        Sends notification with the provided message.
         """
 
+    @notify(NotificationEvent.APPLY_START)
+    def notify_apply_start(self, environment: str) -> None:
+        """Notify when an apply starts."""
+        self.send(NotificationStatus.INFO, f"Plan apply started for environment `{environment}`.")
+
+    @notify(NotificationEvent.RUN_START)
+    def notify_run_start(self, environment: str) -> None:
+        """Notify when an apply starts."""
+        self.send(NotificationStatus.INFO, f"SQLMesh run started for environment `{environment}`.")
+
+    @notify(NotificationEvent.APPLY_FAILURE)
+    def notify_apply_failure(self, exc: Exception) -> None:
+        """Notify in the case of an apply failure."""
+        self.send(NotificationStatus.FAILURE, f"Failed to apply plan.\n{exc}")
+
+    @notify(NotificationEvent.RUN_FAILURE)
+    def notify_run_failure(self, exc: Exception) -> None:
+        """Notify in the case of a run failure."""
+        self.send(NotificationStatus.FAILURE, "Failed to run SQLMesh.\n{exc}")
+
+    @notify(NotificationEvent.AUDIT_FAILURE)
+    def notify_audit_failure(self, audit_error: AuditError) -> None:
+        """Notify in the case of an audit failure."""
+        self.send(NotificationStatus.FAILURE, str(audit_error))
+
+    @property
+    def is_configured(self) -> bool:
+        return True
+
+
+class NotificationTargetManager:
+    """Wrapper around a list of notification targets.
+
+    Calling a notification target's "notify_" method on this object will call it
+    on all registered notification targets.
+    """
+
+    def __init__(
+        self,
+        notification_targets: dict[NotificationEvent, set[BaseNotificationTarget]] | None = None,
+        user_notification_targets: dict[str, set[BaseNotificationTarget]] | None = None,
+        username: str = "",
```

Comment:
[Nit] I think a more standard pattern in the codebase is to make it optional and default to None instead of empty string?